### DISCUSSION TITLE
custom keybinding question

### DISCUSSION THREAD

**Discussion by 0x8f701** [ghostty-org/ghostty#3431] [Q&A] [answered]
I need to move the following keybinding to ghostty, how to do that?

```
[[keyboard.bindings]]
chars = "\u001B0"
key = "Key0"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B1"
key = "Key1"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B2"
key = "Key2"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B3"
key = "Key3"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B4"
key = "Key4"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B5"
key = "Key5"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B6"
key = "Key6"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B7"
key = "Key7"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B8"
key = "Key8"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B9"
key = "Key9"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B."
key = "Period"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001B="
key = "Equals"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001Bf"
key = "F"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001Bb"
key = "B"
mods = "Alt"

[[keyboard.bindings]]
chars = "\u001Bo"
key = "O"
mods = "Alt"

```

**Comment by 0x8f701** [comment]
Tried the following:

```
keybind alt+0 = text:\x1b0
keybind alt+1 = text:\x1b1
keybind alt+2 = text:\x1b2
keybind alt+3 = text:\x1b3
keybind alt+4 = text:\x1b4
keybind alt+5 = text:\x1b5
keybind alt+6 = text:\x1b6
keybind alt+7 = text:\x1b7
keybind alt+8 = text:\x1b8
keybind alt+9 = text:\x1b9
keybind alt+. = text:\x1b.
keybind alt+= = text:\x1b=
keybind alt+f = text:\x1bf
keybind alt+b = text:\x1bb
keybind alt+o = text:\x1bo

``

**Comment by qwerasd205** [comment]
Questions go in discussions. Additionally, please [read the docs](https://ghostty.org/docs/config/keybind)

**Comment by 0x8f701** [comment]
@qwerasd205 could you help moving this into discussion? yes I read the docs

**Reply by qwerasd205** [reply]
> Ghostty supports flexible, custom keybindings through the [keybind configuration option](https://ghostty.org/docs/config/reference#keybind). The configuration reference has detailed reference documentation for the option, but this page will give a high-level overview of how to specify keybindings in Ghostty.
> 
> The basic syntax of a keybinding configuration is:
> ```ini
> keybind = trigger=action
> ```
You are missing the first `=`.

**Reply by 0x8f701** [reply]
```
font-family = "Hack Nerd Font"
font-family-bold = "Hack Nerd Font"
font-family-italic = "Hack Nerd Font"
font-family-bold-italic = "Hack Nerd Font"
shell-integration = fish
font-thicken = true
keybind = alt+0 = text:\x1b0
keybind = alt+1 = text:\x1b1
keybind = alt+2 = text:\x1b2
keybind = alt+3 = text:\x1b3
keybind = alt+4 = text:\x1b4
keybind = alt+5 = text:\x1b5
keybind = alt+6 = text:\x1b6
keybind = alt+7 = text:\x1b7
keybind = alt+8 = text:\x1b8
keybind = alt+9 = text:\x1b9
keybind = alt+. = text:\x1b.
keybind = alt+= = text:\x1b=
keybind = alt+f = text:\x1bf
keybind = alt+b = text:\x1bb
keybind = alt+o = text:\x1bo
```

I tried this it didn't work either

**Reply by rhodes-b** [reply]
@GopherJ for number keys you need to refer to their text name so `alt+zero`, `alt+one` 

similar for `period` and `equal` I think they may work as literals but I am not confident

> The set of possible keys is currently only available for reference by [viewing the source code](https://github.com/ghostty-org/ghostty/blob/d6e76858164d52cff460fedc61ddf2e560912d71/src/input/key.zig#L255). This list is derived from Chromium's keycodes and is closely associated with keys that have associated USB HID codes.

this is quoted in the docs but is easy to miss
the docs also make it seem like you shouldnt add spaces for the second equal sign although im not sure if this is an actual issue
(just tested it threw an error when I had a space so dont do that)

an example for one of your keybinds which worked for me would be
`keybind = alt+one=text:\x1b0`

**Reply by rhodes-b** [reply]
![image](https://github.com/user-attachments/assets/6a741a62-b0bf-43cc-97f6-1cf444ee260a)
 shows it is bound for me you can run
 `ghostty +list-keybinds` to show your keybinds (along with some defaults)

**Reply by gonstoll** [reply]
Side question (sorry to interrupt): I've been pulling my hair off trying to understand how to map those escape sequences to actual keys. Any secret you care to share for knowing what a key combination means in escape sequence? Like, how do I know that `Ctrl-f` is equivalent to `\x06`? So far, the only thing I've found that actually helped was asking chatgpt, and it was not easy to get it to tell me what I wanted :/

**Reply by 00-kat** [reply]
@gonstoll quite late but, A is the first letter so <kbd>ctrl</kbd><kbd>a</kbd> is `\x01`, B is the second letter so <kbd>ctrl</kbd><kbd>b</kbd> is `\x02`, etc.  It's in hex, so <kbd>ctrl</kbd><kbd>z</kbd> would be `\x1a` and not `\x26`.  <kbd>alt</kbd><kbd>\<key\></kbd> is encoded as `\x1b<key>`.  For some of the other ones—I have no idea; but there exists tools which show you what keys it receives, such as `kitten show-key` (ships as a part of Kitty or you can get standalone `kitten` binaries from their GitHub) which print out the encoded form for whatever key you press.

**Comment by absorpheus** [comment]
@0x8f701

The basic syntax of a keybinding configuration is:

```keybind = trigger=action```

Source: https://ghostty.org/docs/config/keybind

Try changing it to:

```
keybind = alt+0=text:\x1b0
keybind = alt+1=text:\x1b1
keybind = alt+2=text:\x1b2
keybind = alt+3=text:\x1b3
keybind = alt+4=text:\x1b4
keybind = alt+5=text:\x1b5
keybind = alt+6=text:\x1b6
keybind = alt+7=text:\x1b7
keybind = alt+8=text:\x1b8
keybind = alt+9=text:\x1b9
```
If that works then you can easily add the other keybinds.

**Comment by 00-kat** [comment, marked as answer]
What is the point of those anyway?  That just rebinds <kbd>alt</kbd><kbd>0</kbd> to send <kbd>alt</kbd><kbd>0</kbd>, <kbd>alt</kbd><kbd>1</kbd> to send <kbd>alt</kbd><kbd>1</kbd>, etc.  Perhaps you're looking to unbind them, i.e. `keybind = alt+one=unbind`.  Or maybe your layout has it in a different position or something, in which case you would want `keybind = alt+physical:one=text:\x1b1`.
